mm2src/coins/sia/src/transaction.rs

```diff
@@ -0,0 +1,1587 @@
+#[derive(Clone, Debug)]
+pub enum SiacoinOutputVersion<'a> {
+    V1(&'a SiacoinOutput),
+    V2(&'a SiacoinOutput),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct SiacoinOutput {
+    pub value: Currency,
+    pub address: Address,
+}
+
+impl<'a> Encodable for SiacoinOutputVersion<'a> {
+    fn encode(&self, encoder: &mut Encoder) {
+        match self {
+            SiacoinOutputVersion::V1(v1) => {
+                CurrencyVersion::V1(&v1.value).encode(encoder);
+                v1.address.encode(encoder);
+            },
+            SiacoinOutputVersion::V2(v2) => {
+                CurrencyVersion::V2(&v2.value).encode(encoder);
+                v2.address.encode(encoder);
+            },
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct CoveredFields {
+    pub whole_transaction: bool,
+    pub siacoin_inputs: Vec<u64>,
+    pub siacoin_outputs: Vec<u64>,
+    pub file_contracts: Vec<u64>,
+    pub file_contract_revisions: Vec<u64>,
+    pub storage_proofs: Vec<u64>,
+    pub siafund_inputs: Vec<u64>,
+    pub siafund_outputs: Vec<u64>,
+    pub miner_fees: Vec<u64>,
+    pub arbitrary_data: Vec<u64>,
+    pub signatures: Vec<u64>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TransactionSignature {
+    pub parent_id: H256,
+    pub public_key_index: u64,
+    pub timelock: u64,
+    pub covered_fields: CoveredFields,
+    pub signature: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct FileContract {
+    pub filesize: u64,
+    pub file_merkle_root: H256,
+    pub window_start: u64,
+    pub window_end: u64,
+    pub payout: Currency,
+    pub valid_proof_outputs: Vec<SiacoinOutput>,
+    pub missed_proof_outputs: Vec<SiacoinOutput>,
+    pub unlock_hash: H256,
+    pub revision_number: u64,
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct V2FileContract {
+    pub filesize: u64,
+    #[serde_as(as = "FromInto<PrefixedH256>")]
+    pub file_merkle_root: H256,
+    pub proof_height: u64,
+    pub expiration_height: u64,
+    pub renter_output: SiacoinOutput,
+    pub host_output: SiacoinOutput,
+    pub missed_host_value: Currency,
+    pub total_collateral: Currency,
+    #[serde_as(as = "FromInto<PrefixedPublicKey>")]
+    pub renter_public_key: PublicKey,
+    #[serde_as(as = "FromInto<PrefixedPublicKey>")]
+    pub host_public_key: PublicKey,
+    pub revision_number: u64,
+    #[serde_as(as = "FromInto<PrefixedSignature>")]
+    pub renter_signature: Signature,
+    #[serde_as(as = "FromInto<PrefixedSignature>")]
+    pub host_signature: Signature,
+}
+
+impl V2FileContract {
+    pub fn with_nil_sigs(&self) -> V2FileContract {
+        debug_assert!(
+            Signature::from_bytes(&[0u8; 64]).is_ok(),
+            "nil signature is valid and cannot return Err"
+        );
```

Comment:
do we actually need this `debug_assert!`? we already use expect below which will panic in either debug or release mode.